### ISSUE TITLE
Tag versioning

### ISSUE THREAD

**Issue by esavier** [isocpp/CppCoreGuidelines#684]
Hi
I would like to create document that would link to version of guidelines, hovewer currently only master version is present and its changing rather dynamically just over one month and i cant link to dynamic content. I considered linking to blob version (latest commit) but this is not how it should be done. 
Would you consider creating tags that would hold static version of guidelines?

**Comment by BjarneStroustrup** [comment]
Doesn't the anchors, the embedded URLs (e.g., https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#nl5-dont-encode-type-information-in-names ) do that? They don't change often; some have not changed since day 1.

**Comment by esavier** [comment]
Hi Bjarne!
This solution introduces numerous issues, especially while there are changes introduced by contributors or author himself.

Those links are tightly bound to name of rule, and those ought to change sometimes. Its enough to change one character and link old link wont work anymore, 

Another think is that in publications we need static content that cant be changed over time, i.e. wherever there is change in this document, my publication would need to be reviewed against introduced changes again, which is not possible in PDF documents (without regenerating, and resending document again.)

That's why I propose to create something like snapshot representing latest release that we can hard link to in a static (generated, f.e. PDF) document,.

Alternatively, a closed branch mechanism can be introduced - wherever master enters stable/finished  state, a closed copy branch is created. However, from practice i believe that tagging is the best option for this kind of problem.

**Comment by BjarneStroustrup** [comment]
We are well aware that what we have is not perfect, but at least we have something until we have the ideas and time to do something better.

**Comment by serwap** [comment]
Hi Bjarne, 

**we only would like that you create a tag in git** so that we can refer to that tag (and not to a changing head version). Imagine that at the beginning of the project we start developing against a head and then your rules change during the project - this is like chasing a moving target. 

Of course we can refer to the last git push, but then we have no reason to choose any specific revisions and it is anyway to hard to track it with a url. 

To sum up: is it ok to create a tag for the current version of the core guidelines in git? It takes just one git command (git tag) - https://git-scm.com/book/en/v2/Git-Basics-Tagging

Thanks in advance.
Kind regards
Piotr

**Comment by BjarneStroustrup** [comment]
We are not ready to offer “stable releases”. We have a choice between continuously updating the guidelines and maintaining support for a series of versions. So far, we have chosen the former while making sure that rules and numbers of rules rarely change. We change many details of rules and add new rules, but this is not a language standard or a library implementation where users have to worry about subtle difference in the meaning of “features.” We do not recycle rule numbers (we did in one case and decided it was a mistake never to be repeated).
We don’t yet know which criteria we would use to determine what constitutes a release and defining releases would encourage forking when people define their own organizational/corporate/personal set of guidelines with additions and subtractions to the core.
As ever, you can always check out or reference any version of the guidelines at any single point in the Git timeline, at a specific date (second line of the document), or at a specific commit.

**Comment by tkruse** [comment]
It may be useful for people unfamiliar with github to know that the first 6+ chars of the sha-ID of a commit are sufficient to link to a fixed commit version:

https://raw.githubusercontent.com/isocpp/CppCoreGuidelines/4d7b1d7b/CppCoreGuidelines.md (August 28, 2016)


Also it might be a temporary compromise to tag the versions that changed the date (second line of the document) with a tagname that merely also reflects that same date, e.g. 

4d7b1d7b prealpha-2016-08-28

giving a more meaningful and self-explaining tag (and url) for the given version, at little cost. Additionally a strictly increasing version number would indicate a hint of the amount of activity that happened between dates:

4d7b1d7b prealpha-**001**-2016-08-28

Note also that any tagging can independently be done in any other fork of this repository.

Just my 5 cents, no response necessary.

**Comment by MikeGitb** [comment]
@BjarneStroustrup 
> We have a choice between continuously updating the guidelines and maintaining support for a series of versions.

Correct me if I'm wrong, but that sounds like you are thinking in terms of independent release branches (like gcc-5.x.y and gcc6.x.y) that get fixes (like updated examples or fixed spelling mistakes) but no new Rules . I might be mistaken, but I believe that is not at all what esavier and other's had in mind. I think it is really just about giving a certain commit a more friendly name than the comit's hash, which makes it somewhat easier to access, communicate and distinguish them for humans and in general would indicate that it is not some "broken" (whatever that means for such a document) intermediate commit.

Personally I don't really care much one way or the other, but tkruse's suggestion sounds viable. Although I'd probably rate limit this somewhat to maybe one tag per month or so (I don't know how often this second line is changed currently) - just so that the output of `git tag` doesn't get overrun by mostly meaningless entries.

**Comment by BjarneStroustrup** [comment]
Please help me understand: exactly what problem are we trying to solve?

**Comment by serwap** [comment]
@BjarneStroustrup our problem is:


# even if you don't reuse the rule ids for a different purpose, you are modifying them. This is fine, as you are not creating a new language standard, but it is rather an on-going work on a new document

# however, various teams would like to have a hard url link to a concete version, so that such guidelines can be imposed in projects (note: in addition to other testable guidelines like hic or misra)

# so we cannot just say: please refer to "core cpp github latest version", because projects may take years to complete. So at the beginning, your rules are different than at the end of a project. So if we do e.g. code inspection against latest version and the code is written e.g. 2 years ago, we have high chances to find violations

# we are creating also some C++ rules (for safety, it has a quite different purpose than your rules) and we would like to refer to various existing standards (e.g. hic, cert, misra). While those documents have a frozen version (e.g. hic 4.0, misra 2008 rule xyz), we cannot refer to your rules (which are very good btw)  as the contents evolve.

To resolve the problem we can of course refer to a random latest version from git (e.g.  from today), but it would be somehow nicer that you could just label/tag it, so that teams could more easily agree on using  that version.

I hope my clarifications somehow help.

Regards
Piotr

**Comment by rianquinn** [comment]
If I might add, having tagged versions would give the analysis tools a "tag" to shoot for in a given release. For example, can anyone easily determine which rules are / are not covered by Clang Tidy? Right now, the only way to do that (that I know) is to take the current version of Clang Tidy, and HEAD from the GitHub page for the Core Guidelines and manually diff. If there was a tagged version of the core guidelines, future versions of Clang Tidy could focus on a tag for support. 

In addition, users of the Core Guidelines could badge their GitHub pages someday with support for a specific Core Guidelines tag (badging with a commit hash would be pretty useless).

**Comment by MikeGitb** [comment]
@serwap: No need to shout ;)
The following part is not quite clear to me:

> To resolve the problem we can of course refer to a random latest version from git (e.g. from today), but it would be somehow nicer that you could just label/tag it, so that teams could more easily agree on using that version.  

What makes a tag so much nicer/how does it ease the decision process? You can after all (as explained by tkruse) already have a  hard link to e.g.

https://github.com/isocpp/CppCoreGuidelines/blob/3ca23f1/CppCoreGuidelines.md#f20-for-out-output-values-prefer-return-values-to-output-parameters

all a (plain) tag would imho do is change that to e.g. 

https://github.com/isocpp/CppCoreGuidelines/blob/prealpha-2016-08-30/CppCoreGuidelines.md#f20-for-out-output-values-prefer-return-values-to-output-parameters

The only advantage I see, is that it is easier to see which of two tags is newer (assuming a proper naming scheme). Or is it about shifting the "selection process" to the maintainers of this repo? What should be the basis of such a selection anyway, except using the most recent available version. For anything more sophisticated, we are well beyond simple tags and almost-zero-overhead.

**Comment by tkruse** [comment]
Creating one such pre-alpha release tag for a suitable-looking commit of this summer and then not talk about tags anymore until summer 2017 might be satisfactory for all parties involved.

Tagging every time the document date is updated was just an initial idea.

**Comment by tkruse** [comment]
I had not noticed before...

In the introductory text, there is already a sentence that does mention versioning: "Had it been an open source (code) project, this would have been release 0.7".

It was recently updated from 0.6 in 1ddd9cdb7abc18. Before that, it had been updated in f2a900b542 (Sep 2015) from 0.1 to 0.6.

So there was some "soft" releasing going on anyway, in the minds of editors.

So by tagging 
* f2a900b542 as "0.6"or "prealpha-0.6-2015-09-14"
* 1ddd9cdb7abc18 as "0.7"or "prealpha-0.7-2016-08-24"

no additional information, workflow or process is introduced that was not present until now.

**Comment by esavier** [comment]
> What makes a tag so much nicer/how does it ease the decision process? You can after all (as explained by tkruse) already have a hard link to...
> @MikeGitb 

ok, 
By release tags we meant something like one tag over 6mths/1year at very best. 
We just cant take any arbitrary commit because people can throw accusation that we have chosen this not other release because it supports our claims.  With tags we can say 'Hey guys, this is what we were working on, it is pointing to latest snapshot of CPPCoreGuideines and it staking to consideration this specific version'

Also as you pointed out, we could distinguish newer and/or older tags if we will be able to release two or more consecutive documents.

**Comment by AndrewPardoe** [comment]
As TKruse notes, Bjarne periodically updates the version number and you can get to those versions directly through git tags: 

f2a900b as "0.6"or "prealpha-0.6-2015-09-14"
1ddd9cd as "0.7"or "prealpha-0.7-2016-08-24"

I will create an index to these.

**Comment by AndrewPardoe** [comment]
The README is updated as well as https://github.com/isocpp/CppCoreGuidelines/releases. Thanks, everyone, for your input and advice.
